src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlDocletWriter.java

```diff
@@ -1492,17 +1499,31 @@ private void generateHeadingId(StartElementTree node, List<? extends DocTree> tr
             } else if (docTree instanceof LinkTree link) {
                 var label = link.getLabel();
                 sb.append(label.isEmpty() ? link.getReference().getSignature() : label.toString());
+            } else if (id == null && docTree instanceof StartElementTree nested
+                    && equalsIgnoreCase(nested.getName(), "a")) {
+                // Use id of embedded anchor element if present
+                id = getIdAttributeValue(nested).orElse(null);
             } else if (docTree instanceof EndElementTree endElement
                     && equalsIgnoreCase(endElement.getName(), tagName)) {
                 break;
-            } else if (docTree instanceof StartElementTree nested
-                    && equalsIgnoreCase(nested.getName(), "a")
-                    && hasIdAttribute(nested)) {
-                return; // Avoid generating id if embedded <a id=...> is present
             }
         }
-        HtmlId htmlId = htmlIds.forHeading(sb, headingIds);
-        content.add("id=\"").add(htmlId.name()).add("\"");
+        String headingContent = sb.toString().trim();
+        if (id == null) {
+            // Generate id attribute
+            HtmlId htmlId = htmlIds.forHeading(headingContent, headingIds);
+            id = htmlId.name();
+            attrs.add("id=\"").add(htmlId.name()).add("\"");
```

Comment:
I'm not saying this is wrong, but it does look like a weird way to achieve the goal.  It seems like either the wrong type is being passed in (`Content` instead of `HtmlTree`) or we're missing some abstraction here.